api/src/main/java/io/opencensus/trace/Tracestate.java

```diff
@@ -0,0 +1,239 @@
+  private static final int KEY_MAX_SIZE = 256;
+  private static final int VALUE_MAX_SIZE = 256;
+  private static final int MAX_KEY_VALUE_PAIRS = 32;
+
+  public static final Tracestate EMPTY = Tracestate.create(new ArrayList<Entry>());
+
+  /**
+   * Returns the value to which the specified key is mapped, or null if this map contains no mapping
+   * for the key.
+   *
+   * @param key with which the specified value is to be associated
+   * @return the value to which the specified key is mapped, or null if this map contains no mapping
+   *     for the key.
+   * @since 0.16
+   */
+  @Nullable
+  public String get(String key) {
+    for (Entry entry : getEntries()) {
+      if (entry.getKey().equals(key)) {
+        return entry.getValue();
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Returns a {@link List} view of the mappings contained in this {@code TraceState}.
+   *
+   * @return a {@link List} view of the mappings contained in this {@code TraceState}.
+   * @since 0.16
+   */
+  public abstract List<Entry> getEntries();
+
+  /**
+   * Returns a builder based on this {@code Tracestate}.
+   *
+   * @return a builder based on this {@code Tracestate}.
+   * @since 0.16
+   */
+  public Builder toBuilder() {
+    return new Builder(this);
+  }
+
+  /**
+   * Builder class for {@link MessageEvent}.
+   *
+   * @since 0.16
+   */
+  public static final class Builder {
+    private final Tracestate parent;
+    private List<Entry> toAdd;
+
+    private Builder(Tracestate parent) {
+      this.parent = parent;
+    }
+
+    /**
+     * Adds a key-value pair.
+     *
+     * @param key the key for the {@code Entry}.
+     * @param value the value for the {@code Entry}.
+     * @return this.
+     */
+    public Builder add(String key, String value) {
+      if (toAdd == null) {
+        toAdd = new ArrayList<Entry>();
+      }
+      // Inserts the element at the front of this list.
+      toAdd.add(0, Entry.create(key, value));
+      return this;
+    }
+
+    /**
+     * Builds a TraceState by adding the entries to the parent in front of the key-value pairs list
+     * and removing duplicate entries.
+     *
+     * @return a TraceState with the new entries.
+     */
+    public Tracestate build() {
+      return create(parent, toAdd);
+    }
+
+    private static Tracestate create(Tracestate parent, List<Entry> toAdd) {
+      Utils.checkNotNull(parent, "parent");
+      if (toAdd == null || toAdd.size() == 0) {
+        return parent;
+      }
+
+      List<Entry> ret = new ArrayList<Entry>(toAdd.size() + parent.getEntries().size());
+      ret.addAll(toAdd);
+      for (Entry entry : parent.getEntries()) {
+        boolean isPresent = false;
+        for (Entry entryExtra : toAdd) {
+          if (entryExtra.getKey().equals(entry.getKey())) {
+            isPresent = true;
+            break;
+          }
+        }
+        if (!isPresent) {
+          ret.add(entry);
+        }
+      }
+
+      return Tracestate.create(ret);
```

Comment:
overall length validation should probably happen here

Reply:
It happens in create to make sure that any caller will pass the check.

Reply:
I mean there is an overall header size limit of 512 bytes:

> The length of a combined header MUST be less than or equal to 512 bytes. If the length of a combined header is more than 512 bytes it SHOULD be ignored.

I think we should try to follow the protocol and check if it's too hard to implement or too inefficient, it's a good moment to fix the protocol.